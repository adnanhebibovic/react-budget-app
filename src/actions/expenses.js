import firestore from '../services/firebase'
import {v4 as uuid} from 'uuid'

export const addExpenseToStore = function(expense = {}) {
    const defaults = {
        id: new uuid(),
        title: '',
        amount: 0,
        createdAt: new Date().getTime(),
        ...expense
    };

    return {
        type: 'ADD_EXPENSE',
        expense: {
            id: defaults.id,
            title: defaults.title,
            amount: defaults.amount,
            createdAt: defaults.createdAt
        }
    };
}

export const setExpensesInStore = function(expenses = []) {
    return {
        type: 'SET_EXPENSE',
        expenses
    }
}

export const removeExpenseFromStore = function(id) {
    return {
        type: 'REMOVE_EXPENSE',
        id
    }
}

export const editExpenseInStore = function(id, expense) {
    const defaults = {
        title: '',
        amount: 0,
        ...expense
    };

    return {
        type: 'EDIT_EXPENSE',
        id,
        expense: defaults
    }
}

export const startAddExpense = function(uid, expense = {}) {
    const defaults = {
        createdAt: new Date().getTime(),
        ...expense
    }

    return function(dispatch) {
        const user = firestore.collection('users').doc(uid)
        return user.collection('expenses').add(defaults)
        .then((doc) => {
            dispatch(addExpenseToStore({
                id: doc.id,
                ...defaults
            }))
        })
    }
}

export const startRemoveExpense = function(uid, id) {
    return function(dispatch) {
        const user = firestore.collection('users').doc(uid)
        return user.collection('expenses').doc(id).delete()
        .then(() => {
            dispatch(removeExpenseFromStore(id))
        })
    }
}

export const startEditExpense = function(uid, id, expense) {
    return function(dispatch) {
        const user = firestore.collection('users').doc(uid)
        return user.collection('expenses').doc(id).update(expense)
        .then(() => {
            dispatch(editExpenseInStore(id, expense))
        })
    }
}

export const startSetExpense = function(uid) {
    return function(dispatch) {
        const user = firestore.collection('users').doc(uid)
        return user.collection('expenses').get()
        .then((snapshot) => {
            const expenses = []
            snapshot.forEach((document) => {
                expenses.push({
                    id: document.id, 
                    ...document.data()
                })
            })
            return expenses;
        }).then((result) => {
            dispatch(setExpensesInStore(result))
        })
    }
}