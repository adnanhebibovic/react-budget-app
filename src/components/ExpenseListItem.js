import React from 'react'
import Moment from 'react-moment';
import { NavLink } from 'react-router-dom'
import numeral from 'numeral'

function ExpenseListItem({id, title, amount, createdAt}) {
    return (
        <NavLink className="list-item" to={`/expenses/${id}`}>
            <div>
                <h3 className="list-item__title">{title}</h3>
                <span className="list-item__date"><Moment format="MMMM, Do YYYY">{createdAt}</Moment></span>
            </div>
            <h3 className="list-item__amount">{numeral(amount).format('$0,0[.]00')}</h3>
        </NavLink>
    )
}

export default ExpenseListItem 