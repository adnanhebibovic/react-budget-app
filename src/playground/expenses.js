import { createStore, combineReducers } from 'redux'
import uuid from 'uuid'

const addExpense = function({title = '', amount = 0} = {}) {
    return {
        type: 'ADD_EXPENSE',
        expense: {
            id: uuid(),
            title,
            amount,
            createdAt: new Date()
        }
    }
}

const removeExpense = function(id = uuid()) {
    return {
        type: 'REMOVE_EXPENSE',
        id
    }
}

const editExpense = function(id = uuid(), expense = {}) {
    return {
        type: 'EDIT_EXPENSE',
        id,
        expense
    }
}

const expensesDefaultState = []

const expensesReducer = function(state = expensesDefaultState, action) {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return state.concat(action.expense)
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id)
        case 'EDIT_EXPENSE':
            return state.map((expense) => expense.id == action.id 
                ? {...expense, ...action.expense}
                : expense
            )
        default:
            return state;
    }
}

const setTextFilter = function(text = undefined) {
    return {
        type: 'SET_TEXT_FILTER',
        text
    }
}

const sortByAmount = function() {
    return {
        type: 'SORT_BY',
        sortBy: 'amount'
    }
}

const sortByDate = function() {
    return {
        type: 'SORT_BY',
        sortBy: 'createdAt'
    }
}

const setDateFilters = function({startDate = undefined, endDate = undefined} = {}) {
    return {
        type: 'SET_DATE_FILTERS',
        dates: {
            startDate,
            endDate
        }
    }
}

const filtersDefaultState = {
    text: undefined,
    sortBy: undefined,
    startDate: undefined,
    endDate: undefined
}

const filtersReducer = function(state = filtersDefaultState, action) {
    switch (action.type) {
        case 'SET_TEXT_FILTER':
            return {
                ...state,
                text: action.text
            }
        case 'SORT_BY':
            return {
                ...state,
                sortBy: action.sortBy
            }
        case 'SET_DATE_FILTERS':
            return {
                ...state,
                ...action.dates
            }
        default:
            return state;
    }
}

const getExpenses = function(expenses, {text, sortBy, startDate, endDate}) {
    const result = expenses.filter((expense) => {
        const textMatch = typeof text === 'undefined' || expense.title.includes(text);
        const startDateMatch = typeof startDate === 'undefined' || expense.createdAt >= startDate;
        const endDateMatch = typeof endDate === 'undefined' || expense.createdAt <= endDate;

        return textMatch && startDateMatch && endDateMatch;
    })

    if (sortBy)
    {
        return result.sort((e1, e2) => {
            if (e1[sortBy] < e2[sortBy]) {
                return -1;
            } else if (e1[sortBy] > e2[sortBy]) {
                return 1;
            } else {
                return 0;
            }
        })
    }

    return result
}

const store = createStore(
    combineReducers({
        expenses: expensesReducer,
        filters: filtersReducer
    })
)

store.dispatch(addExpense({
    title: 'January bills',
    amount: 279.00 
}))

store.dispatch(addExpense({
    title: 'February bills',
    amount: 523.00 
}))

store.dispatch(addExpense({
    title: 'March bills',
    amount: 358.00 
}))

store.dispatch(addExpense({
    title: 'New car',
    amount: 1000.00 
}))

store.subscribe(() => {
    const state = store.getState();
    console.log(getExpenses(state.expenses, state.filters))
})

store.dispatch(setTextFilter('bills'))
store.dispatch(sortByAmount())
store.dispatch(sortByDate())
store.dispatch(setDateFilters({endDate: new Date()}))
