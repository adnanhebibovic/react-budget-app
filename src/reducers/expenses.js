export const expensesDefaultState = []

export default function(state = expensesDefaultState, action) {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return state.concat(action.expense)
        case 'SET_EXPENSE':
            return state.concat(action.expenses.filter((expense) => state.findIndex((item) => item.id === expense.id) < 0))
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id)
        case 'EDIT_EXPENSE':
            return state.map((expense) => expense.id === action.id 
                ? {...expense, ...action.expense}
                : expense
            )
        default:
            return state;
    }
}