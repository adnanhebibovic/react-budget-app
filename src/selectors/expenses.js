export default function(expenses, {text, sortBy, startDate, endDate}) {
    const result = expenses.filter((expense) => {
        const textMatch = typeof text == 'undefined' || expense.title.toLowerCase().includes(text.toLowerCase());
        const startDateMatch = typeof startDate == 'undefined' || expense.createdAt > startDate;
        const endDateMatch = typeof endDate == 'undefined' || expense.createdAt < endDate;

        return textMatch && startDateMatch && endDateMatch;
    })

    if (sortBy) {
        return result.sort((e1, e2) => {
            if (e1[sortBy] < e2[sortBy]) {
                return -1;
            } else if (e1[sortBy] > e2[sortBy]) {
                return 1;
            } else {
                return 0;
            }
        })
    }

    return result;
}
