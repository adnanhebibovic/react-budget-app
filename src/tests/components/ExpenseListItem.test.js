import React from 'react'
import {BrowserRouter} from 'react-router-dom'
import { create, act } from 'react-test-renderer';

import ExpenseListItem from '../../components/ExpenseListItem'
import expenses from '../fixtures/expenses'

test('Should render ExpenseListItem with expenses', () => {
    let root
    act(() => {
        root = create(<BrowserRouter><ExpenseListItem {...expenses[0]}></ExpenseListItem></BrowserRouter>)
    });
    expect(root).toMatchSnapshot()
})
