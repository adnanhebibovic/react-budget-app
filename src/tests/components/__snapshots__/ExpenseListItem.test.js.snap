// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseListItem with expenses 1`] = `
<div>
  <a
    aria-current={null}
    href="/expenses/1"
    onClick={[Function]}
  >
    <h3>
      first expense
    </h3>
  </a>
  <p>
    $400
  </p>
  <time
    dateTime={"2015-03-24T23:00:00.000Z"}
  >
    Wed Mar 25 2015 00:00:00 GMT+0100
  </time>
</div>
`;
